### DISCUSSION TITLE
Custom server + getServerSideProps

### DISCUSSION THREAD

**Discussion by anton-johansson** [vercel/next.js#23144] [Help] [answered]
I'm using a custom server implementation where I wire up a lot of stuff, such as database connections and similar. I need to access some of these dependencies in my `getServerSideProps` hook, but I can't figure out how to do it. I've tried some "nasty" workarounds, where I statically set a dependency:
```ts
let instance: SessionManager;
export const getSessionManager = () => {
    return instance;
};

export const setSessionManager = (sessionManager: SessionManager) => {
    instance = sessionManager;
};
```

I use `setSessionManager` from my custom server where I wire up all dependencies, and I use `getSessionManager` in my `getServerSideProps` hook, but it is always `undefined` in that situation. It's like `getServerSideProps` is executed completely separate from the custom server.

Does anyone have any suggestions on how I can do stuff like this?

Thanks in advance!

**Comment by akellbl4** [comment, marked as answer]
If I understood rightly, a better way to do so is to set your dependency in `req` on a custom server level. Then you can get that form `ctx.req` in `getServerSideProps`. It's also a "messy" way to solve the problem.

`server.js`
```js
app.prepare().then(() => {
  createServer((req, res) => {
    const parsedUrl = new URL(req.url, "http://example.com");
    // set data or methods here
    req.custom = true;
    handle(req, res, parsedUrl);
  }).listen(port, (err) => {
    if (err) throw err;
    console.log(`> Ready on http://localhost:${port}`);
  });
});
```

`page.js`
```js
export function getServerSideProps(ctx) {
  // get them here
  console.log(ctx.req.custom);
  return {
    props: {},
  };
}
```

**Reply by anton-johansson** [reply]
I tried this, and it works perfect for my use case! Thanks a bunch, @akellbl4.

I'm using `koa` as my custom server, and here is what I did:

```ts
const addIOC = (sessionManager: SessionManager, otherServiceToExpose: SomeService) => {
    return async (context: ParameterizedContext, next: Next) => {
        const ioc = {
            sessionManager,
            otherServiceToExpose,
        };

        (context.req as any).ioc = ioc;
        await next();
    };
};

const app = next({dev});
const koa = new KoaRouter();

// ...

koa.all("(.*)", async context => handle(context.req, context.res));
void app.prepare().then(() => {
    const server = new Koa();
    server.use(addIOC(sessionManager, otherServiceToExpose));
    server.use(bodyParser());
    server.use(koa.routes());
    server.listen(port, () => console.log(`Listening on http://localhost:${port}`));
});
```

... and `getServerSideProps`:

```ts
export const getServerSideProps: GetServerSideProps = async ({req, res, query}: GetServerSidePropsContext) => {
    // ...
    const ioc: RequestIOC = (req as any).ioc;
    const sessionManager = ioc.sessionManager;
    // ...

    return {};
};
```

If I ever introduce a real IOC container, I can just pass in the entire container.
